### PR TITLE
Нестеров -- Лабораторная работа #2 

### PR DESCRIPTION
**Pull request by allnes**

### CODE REVIEW

**Comment by allnes** [comment]
@Lenferd @PovelikinRostislav

**Comment by PetrovKP** [comment]
👍

File: code/red-black-tree/View/src/main/java/ru/unn/agile/redblacktree/view/RedBlackTreeView.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    @FXML
+    private Label txtResultRemove;
+
+    @FXML
+    void initialize() {
+
+        txtAddField.textProperty().bindBidirectional(viewModel.addFieldProperty());
+        txtFindField.textProperty().bindBidirectional(viewModel.findInsertFieldProperty());
+        txtRemoveField.textProperty().bindBidirectional(viewModel.removeInsertFieldProperty());
+        txtAddField.textProperty().bindBidirectional(viewModel.addFieldProperty());
```
Comment (Lenferd) [inline code/red-black-tree/View/src/main/java/ru/unn/agile/redblacktree/view/RedBlackTreeView.java]: Same as line 36

File: code/red-black-tree/View/src/main/java/ru/unn/agile/redblacktree/view/RedBlackTreeView.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+            @Override
+            public void handle(final ActionEvent event) {
+                viewModel.addElementToTree();
+            }
+        });
+
+        btnFind.setOnAction(new EventHandler<ActionEvent>() {
+            @Override
+            public void handle(final ActionEvent event) {
+                viewModel.findElementToTree();
```
Comment (Lenferd) [inline code/red-black-tree/View/src/main/java/ru/unn/agile/redblacktree/view/RedBlackTreeView.java]: `findElementInTree`

File: code/red-black-tree/View/src/main/java/ru/unn/agile/redblacktree/view/RedBlackTreeView.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+            }
+        });
+
+        btnRemove.setOnAction(new EventHandler<ActionEvent>() {
+            @Override
+            public void handle(final ActionEvent event) {
+                viewModel.removeElementToTree();
```
Comment (Lenferd) [inline code/red-black-tree/View/src/main/java/ru/unn/agile/redblacktree/view/RedBlackTreeView.java]: `removeElementFromTree`

File: code/red-black-tree/ViewModel/src/test/java/ru/unn/agile/redblacktree/viewmodel/ViewModelTest.java [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+
+    @Test
+    public void canReportBadFormatRemoveInsertField() {
+        viewModel.removeInsertFieldProperty().set("c");
+
+        assertEquals(Status.BAD_FORMAT.toString(), viewModel.statusProperty().get());
+    }
+
+    @Test
+    public void calculateButtonIsDisabledInitiallyForAddElement() {
```
Comment (Lenferd) [inline code/red-black-tree/ViewModel/src/test/java/ru/unn/agile/redblacktree/viewmodel/ViewModelTest.java]: `calculateButton`?

File: code/red-black-tree/ViewModel/src/main/java/ru/unn/agile/redblacktree/viewmodel/ViewModel.java [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+        return findInsertField;
+    }
+    public StringProperty removeInsertFieldProperty() {
+        return removeInsertField;
+    }
+
+    public StringProperty statusProperty() {
+        return status;
+    }
+    public final String getStatus() {
```
Comment (Lenferd) [inline code/red-black-tree/ViewModel/src/main/java/ru/unn/agile/redblacktree/viewmodel/ViewModel.java:113]: Not used. But seems like should, as in tests a lot of `statusProperty().get()` calls

File: code/red-black-tree/ViewModel/src/main/java/ru/unn/agile/redblacktree/viewmodel/ViewModel.java [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+        return status.get();
+    }
+
+    private Status getInputStatus() {
```
Comment (Lenferd) [inline code/red-black-tree/ViewModel/src/main/java/ru/unn/agile/redblacktree/viewmodel/ViewModel.java]: Not a good idea to validate all inputs at once, but at least name it like `getInputsStatus`

File: code/red-black-tree/ViewModel/src/test/java/ru/unn/agile/redblacktree/viewmodel/ViewModelTest.java [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+    }
+
+    @Test
+    public void statusIsWaitingWhenRemoveElementToTreeWithEmptyFields() {
+        viewModel.removeElementToTree();
+        assertEquals(Status.WAITING.toString(), viewModel.statusProperty().get());
+    }
+
+    @Test
+    public void statusIsReadyWhenFieldsAreFill() {
```
Comment (Lenferd) [inline code/red-black-tree/ViewModel/src/test/java/ru/unn/agile/redblacktree/viewmodel/ViewModelTest.java]: It has to be three different tests.

**Comment by allnes** [comment]
@Lenferd corrected

File: code/red-black-tree/View/src/main/java/ru/unn/agile/redblacktree/view/RedBlackTreeView.java (hunk 1 above)
Comment (Lenferd) [reply code/red-black-tree/View/src/main/java/ru/unn/agile/redblacktree/view/RedBlackTreeView.java]: Resolved

File: code/red-black-tree/View/src/main/java/ru/unn/agile/redblacktree/view/RedBlackTreeView.java (hunk 2 above)
Comment (Lenferd) [reply code/red-black-tree/View/src/main/java/ru/unn/agile/redblacktree/view/RedBlackTreeView.java]: Resolved

File: code/red-black-tree/View/src/main/java/ru/unn/agile/redblacktree/view/RedBlackTreeView.java (hunk 3 above)
Comment (Lenferd) [reply code/red-black-tree/View/src/main/java/ru/unn/agile/redblacktree/view/RedBlackTreeView.java]: Resolved

File: code/red-black-tree/ViewModel/src/test/java/ru/unn/agile/redblacktree/viewmodel/ViewModelTest.java (hunk 4 above)
Comment (Lenferd) [reply code/red-black-tree/ViewModel/src/test/java/ru/unn/agile/redblacktree/viewmodel/ViewModelTest.java]: Resolved

File: code/red-black-tree/ViewModel/src/main/java/ru/unn/agile/redblacktree/viewmodel/ViewModel.java (hunk 6 above)
Comment (Lenferd) [reply code/red-black-tree/ViewModel/src/main/java/ru/unn/agile/redblacktree/viewmodel/ViewModel.java]: Resolved

File: code/red-black-tree/ViewModel/src/test/java/ru/unn/agile/redblacktree/viewmodel/ViewModelTest.java (hunk 7 above)
Comment (Lenferd) [reply code/red-black-tree/ViewModel/src/test/java/ru/unn/agile/redblacktree/viewmodel/ViewModelTest.java]: Resolved

File: code/red-black-tree/ViewModel/src/main/java/ru/unn/agile/redblacktree/viewmodel/ViewModel.java [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+
+public class ViewModel {
+    private final StringProperty addField = new SimpleStringProperty();
+    private final StringProperty findInsertField = new SimpleStringProperty();
+    private final StringProperty removeInsertField = new SimpleStringProperty();
+    private RedBlackTree rbTree = new RedBlackTree();
+
+    private final BooleanProperty addElementToTreeDisabled = new SimpleBooleanProperty();
+    private final BooleanProperty findElementToTreeDisabled = new SimpleBooleanProperty();
+    private final BooleanProperty removeElementToTreeDisabled = new SimpleBooleanProperty();
```
Comment (Lenferd) [inline code/red-black-tree/ViewModel/src/main/java/ru/unn/agile/redblacktree/viewmodel/ViewModel.java]: `In`, `From`

File: code/red-black-tree/ViewModel/src/main/java/ru/unn/agile/redblacktree/viewmodel/ViewModel.java (hunk 5 above)
Comment (allnes) [reply code/red-black-tree/ViewModel/src/main/java/ru/unn/agile/redblacktree/viewmodel/ViewModel.java:113]: I corrected this item, but I don't know what could I change?

File: code/red-black-tree/ViewModel/src/main/java/ru/unn/agile/redblacktree/viewmodel/ViewModel.java [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+    }
+    public final boolean isFindElementToTreeDisabled() {
+        return findElementToTreeDisabled.get();
+    }
+    public BooleanProperty removeElementToTreeDisabledProperty() {
+        return removeElementToTreeDisabled;
+    }
+    public final boolean isRemoveElementToTreeDisabled() {
+        return removeElementToTreeDisabled.get();
+    }
```
Comment (Lenferd) [inline code/red-black-tree/ViewModel/src/main/java/ru/unn/agile/redblacktree/viewmodel/ViewModel.java]: In, From

File: code/red-black-tree/ViewModel/src/main/java/ru/unn/agile/redblacktree/viewmodel/ViewModel.java [hunk 10]
```diff
@@ -0,0 +1,204 @@
...
+        BooleanBinding couldRemoveElementToTree = new BooleanBinding() {
+            {
+                super.bind(removeInsertField);
+            }
+            @Override
+            protected boolean computeValue() {
+                return getInputsStatus() == Status.READY;
+            }
+        };
+        removeElementToTreeDisabled.bind(couldRemoveElementToTree.not());
```
Comment (Lenferd) [inline code/red-black-tree/ViewModel/src/main/java/ru/unn/agile/redblacktree/viewmodel/ViewModel.java]: In, From

File: code/red-black-tree/ViewModel/src/main/java/ru/unn/agile/redblacktree/viewmodel/ViewModel.java (hunk 5 above)
Comment (Lenferd) [reply code/red-black-tree/ViewModel/src/main/java/ru/unn/agile/redblacktree/viewmodel/ViewModel.java]: Resolved

**Comment by allnes** [comment]
@Lenferd corrected again

**Review by Lenferd — APPROVED** [review]

**Comment by Lenferd** [comment]
:+1:

**Pull request closed without merging**